Adds various laws to the haddocks

solomon-b opened the pull request:
I added most of the laws to the haddocks. I had trouble coming up with descriptions for about half of them.

Commit:
Adds various laws to the haddocks

kozross commented on src/Test/Monad/Cont.hs:
```diff
@@ -12,36 +12,46 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 callCC_phantom f = callCC f :=: callCC (f . (fmap . fmap) absurd)
 
--- | > callCC (\k -> f k >>= (\x -> k x >>= g k)) = callCC f
+-- | Left identity on callCC
```
Let's keep descriptions to sentences (thus, will full stops). Also, the identifier should be either linked, or written monospace.

kozross commented on src/Test/Monad/Morph.hs:
```diff
@@ -10,12 +10,20 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 
+-- |
+-- @
+-- hom (m '>>=' k) = hom m '>>=' hom . k
```
This needs a natural-language description. Something like 'Natural transformations distribute over `(>>=)`.', perhaps?

Lysxia replied:
Monad morphisms are what distribute over `(>>=)`.

Lysxia replied:
"Commutativity" and "distributivity" often occur interchangeably in this context. "Monad morphisms commute with `(>>=)` and `return`".

kozross commented on src/Test/Monad/Morph.hs:
```diff
@@ -10,12 +10,20 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 
+-- |
+-- @
+-- hom ('return' a) = 'return' a
```
Same as before: something like 'Combining `return` with a natural transformation is the same as using `return` directly.', maybe?

Commit:
Adds more descriptions of laws

solomon-b commented:
I added a bunch more descriptions and updated as per your recommendations. There are still a few that I'm having trouble describing in English, as well as a bunch whose descriptions I don't feel are precise enough.

Commit:
Adds more descriptions and laws

kozross commented:
These all seem good to me (just need full stops). @Lysxia - thoughts or suggestions before I merge?

Lysxia commented:
LGTM

On 9/29/2020 1:23 PM, Koz Ross wrote:
> These all seem good to me (just need full stops). @Lysxia 
> <https://github.com/Lysxia> - thoughts or suggestions before I merge?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/kozross/test-monad-laws/pull/1#issuecomment-700861545>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AATGCALLM3AW3SJ6YM6LWLDSIIJYJANCNFSM4R42CHXQ>.
>

solomon-b commented:
By full stops, do you mean periods to end sentences or something else?

kozross commented:
@ssbothwell I meant exactly periods to end sentences - sorry that I was unclear.

Commit:
Ends all haddock sentences with periods.

solomon-b commented:
Sorry about the delay, i just pushed a change with all the periods added.

Lysxia commented on src/Test/Monad/Cont.hs:
```diff
@@ -12,36 +12,46 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 --
---  > callCC (const x) = x
+-- @
+--  'callCC' ('const' x) = x
```
```suggestion
-- 'callCC' ('const' x) = x
```

Lysxia commented on src/Test/Monad/Cont.hs:
```diff
@@ -12,36 +12,46 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 --
--- > callCC f = callCC (f . (fmap . fmap) absurd)
+-- @
+-- 'callCC' f = 'callCC' (f . ('fmap' . 'fmap') 'absurd')
```
```suggestion
-- 'callCC' f = 'callCC' (\\k -> f ('absurd' '.' k))
-- @
```

Now that I think about it, the pointful version actually seems nicer. You can also change the actual code below (make sure it typechecks!)

Lysxia commented on src/Test/Monad/Cont.hs:
```diff
@@ -12,36 +12,46 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 callCC_phantom f = callCC f :=: callCC (f . (fmap . fmap) absurd)
 
--- | > callCC (\k -> f k >>= (\x -> k x >>= g k)) = callCC f
+-- | The continuation never returns, so g does not matter.
```
```suggestion
-- | The continuation never returns, so @g@ does not matter.
```

Lysxia commented on src/Test/Monad/Control.hs:
```diff
@@ -41,6 +41,10 @@ liftWith_lift
 
+-- |
+-- @
+-- ('liftWith' (\run -> run t) '>>=' 'restoreT' . 'return') = t
```
Backslashes need to be escaped, if I recall correctly. (double check the haddocks)

Lysxia commented on src/Test/Monad/Except.hs:
```diff
@@ -8,24 +8,40 @@ module Test.Monad.Except where
 
+-- | 'catchError' absorbs 'throwError'.
+-- @
+-- 'catchError' ('throwError e') h = h e
```
```suggestion
-- 'catchError' ('throwError' e) h = h e
```

Lysxia commented on src/Test/Monad/Except.hs:
```diff
@@ -35,6 +51,10 @@ catch_catch m h1 h2 =
   :=:
   catchError m (\e -> catchError (h1 e) h2)
 
+-- | Catching from ('return' a) is equivalent to ('return' a).
```
```suggestion
-- | Catching from @('return' a)@ is equivalent to @('return' a)@.
```

Lysxia commented on src/Test/Monad/Morph.hs:
```diff
@@ -10,12 +10,20 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 -- | Natural transformation.
 type m ~> n = forall t. m t -> n t
 
+-- | Monad morphisms commute with (>>=) and (return).
```
```suggestion
-- | Monad morphisms commute with @('>>=')@ and 'return'.
```

Lysxia commented on src/Test/Monad/Morph.hs:
```diff
@@ -10,12 +10,20 @@ import Test.QuickCheck.HigherOrder (Equation(..))
   => (m ~> n) -> m a -> (a -> m b) -> Equation (n b)
 bindHom hom m k = hom (m >>= k) :=: (hom m >>= hom . k)
 
+-- | A natural transformation of ('return' a) is the same as ('return' a).
```
```suggestion
-- | A monad morphism maps @('return' a)@ to @('return' a)@.
```

Lysxia commented on src/Test/Monad/Reader.hs:
```diff
@@ -10,12 +10,20 @@ import Test.QuickCheck.HigherOrder (Equation(..))
   => Equation (m r)
 ask_ask = (ask >> ask) :=: ask @r @m
 
+-- | 'local' can be interchanged with fmap.
```
Not quite. This law works because `ask` is the second argument.

To modify the environment then asking for it is the same as asking for the environment then modifying it.

Lysxia commented on src/Test/Monad/Reader.hs:
```diff
@@ -28,12 +36,20 @@ local_ask f = local f ask :=: fmap @m f ask
 
+-- | 'local' preserves composition.
+-- @
+-- 'local' f ('local' g) = 'local (g . f)'
```
```suggestion
-- 'local' f . 'local' g = 'local (g . f)'
```

Lysxia commented on src/Test/Monad/Reader.hs:
```diff
@@ -28,12 +36,20 @@ local_ask f = local f ask :=: fmap @m f ask
 
+-- | ('local' 'id') is identity on a monadic action.
+-- @
+-- 'local' 'id' m = m
```
```suggestion
-- 'local' 'id' = 'id'
```

Lysxia commented on src/Test/Monad/State.hs:
```diff
@@ -10,15 +10,31 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 
+-- | 'put'/'get' identity. ('put' s '>>' 'get') is the same as ('return' s).
+-- @
+-- 'put s' '>>' 'get' = 'return' s
```
```suggestion
-- 'put' s '>>' 'get' = 'return' s
```

Lysxia commented on src/Test/Monad/State.hs:
```diff
@@ -10,15 +10,31 @@ import Test.QuickCheck.HigherOrder (Equation(..))
 
+-- | 'put' absorption. ('put' s1 '>>' 'put' s2) discards the first 'put'.
+-- @
+-- 'put s1' '>>' 'put s2' = 'put' s2
```
```suggestion
-- 'put' s1 '>>' 'put; s2 = 'put' s2
```

Lysxia commented on src/Test/Monad/Writer.hs:
```diff
@@ -6,24 +6,40 @@ import Data.Functor (($>))
 import Control.Monad.Writer
 import Test.QuickCheck.HigherOrder (Equation(..))
 
+-- | Telling two values sequentially is equivalent to monoidally appending the two values then telling once.
```
```suggestion
-- | Telling two values sequentially is equivalent to appending the two values then telling once.
```

Lysxia commented on src/Test/Monad/Writer.hs:
```diff
@@ -6,24 +6,40 @@ import Data.Functor (($>))
   => w -> w -> Equation (m ())
 tell_tell w1 w2 = (tell w1 >> tell w2) :=: tell (w1 <> w2)
 
+-- | telling a 'mempty' value is equivalent to calling ('return' ()).
```
```suggestion
-- | telling a 'mempty' value is equivalent to calling @('return' ())@.
```

Lysxia commented on src/Test/Monad/Writer.hs:
```diff
@@ -6,24 +6,40 @@ import Data.Functor (($>))
   => Equation (m ())
 tell_mempty = tell mempty :=: return ()
 
+-- | ('listen' . 'return') lifts a value into a monad and tuples it with 'mempty'.
```
```suggestion
-- | @('listen' . 'return')@ lifts a value into a monad and tuples it with 'mempty'.
```

Lysxia commented on src/Test/Monad/Writer.hs:
```diff
@@ -6,24 +6,40 @@ import Data.Functor (($>))
 
+-- |
+-- @
+-- 'listen' (m '>>=' k) = 'listen' m '>>=' \(a, wa) -> 'listen' (k a) >>= \(b, wb) -> 'return' (b, wa '<>' wb)
```
Double check whether the backslaskes are preserved by haddock.
This line is also a bit long.

```suggestion
-- 'listen' (m '>>=' k)
-- = 'listen' m '>>=' \\(a, wa) ->
--   'listen' (k a) '>>=' \\(b, wb) ->
--   'return' (b, wa '<>' wb)
```

Lysxia commented on src/Test/Monad/Writer.hs:
```diff
@@ -34,26 +50,42 @@ listen_bind m k =
     (b, wb) <- listen (k a)
     return (b, wa <> wb)
 
+-- | ('listen' . 'tell') tells a value then lifts it into a monad.
```
```suggestion
-- | @('listen' . 'tell')@ tells a value then lifts it into a monad.
```

Lysxia commented on src/Test/Monad/Writer.hs:
```diff
@@ -34,26 +50,42 @@ listen_bind m k =
 
+-- | ('listen' . 'tell') tells a value then lifts it into a monad.
+-- @
+-- 'listen' ('tell' w) = 'tell' w '>>' 'return' w
```
This seems to be missing the `fmap snd` below.

Lysxia commented:
I wouldn't want to let this linger if you can't get to these last changes. @ssbothwell Would you prefer we merge this PR and keep polishing separately/on another PR?

Commit:
Various haddock corrections

Commit:
Escapes lambdas in haddocks

solomon-b commented:
I just corrected the typos, but I didn't change `(f . ('fmap' . 'fmap') 'absurd')` to be pointful. I skipped it because I'm having trouble building the project currently due to `quickcheck-higherorder` not being in the stable channel for nix. 🤷

Commit:
Additional haddock corrections

Commit:
Updates `local f ask = fmap f ask` description.

Lysxia commented:
No problem, thanks for the help!

solomon-b commented:
Happy to help!

kozross commented:
If @Lysxia is good with merging this, I'll merge.

Lysxia commented:
:+1: @kozross Do you plan to make more changes soon, or should I merge the original PR after this?

kozross commented:
@Lysxia Probably fine to merge - I don't think I could add anything to the work @ssbothwell already did here.

kozross commented:
@ssbothwell Have you addressed all of @Lysxia's comments to their satisfaction? If so, I'll merge.

Lysxia commented:
I think this is good to merge.

On 10/7/2020 5:13 PM, Koz Ross wrote:
> @ssbothwell <https://github.com/ssbothwell> Have you addressed all of 
> @Lysxia <https://github.com/Lysxia>'s comments to their satisfaction? If 
> so, I'll merge.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/kozross/test-monad-laws/pull/1#issuecomment-705196930>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AATGCAK6RTTG6RSF6YA6EALSJTKXVANCNFSM4R42CHXQ>.
>

kozross commented:
And merged! @ssbothwell - thank you so much for doing this!

solomon-b commented:
I didn't have any more changes. Thanks for merging!

On Wed, Oct 7, 2020, 7:39 PM Koz Ross <notifications@github.com> wrote:

> Merged #1 <https://github.com/kozross/test-monad-laws/pull/1> into master.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kozross/test-monad-laws/pull/1#event-3853233790>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF3ZWO5P344KV2FLOEVKTSJUQ7ZANCNFSM4R42CHXQ>
> .
>
